### PR TITLE
Mrp 2020

### PR DESCRIPTION
**Pull request by OfirArviv**

### CODE REVIEW

**Comment by danielhers** [comment]
Thanks, very helpful. Any script updates for training / prediction?

**Comment by OfirArviv** [comment]
> 
> 
> Thanks, very helpful. Any script updates for training / prediction?

It the same really. Don't think there are any changes from the standard ones.
For example for amr:

python -m tupa --seed $RANDOM --cores=16 --use-bert --dynet-gpu --pytorch-gpu --no-validate-oracle --save-every=50000 \
    -t ../mrp/2020/cf/training/amr.mrp -d ../mrp/2020/cf/validation/amr.mrp \
    --conllu ../mrp/2020/cf/companion/udpipe.mrp --alignment ../mrp/2020/cf/companion/isi.mrp -m models/amr -v

File: tupa/oracle.py [hunk 1]
```diff
@@ -193,8 +193,12 @@ def get_node_label(self, state, node):
...
 
     def get_node_property_value(self, state, node):
-        true_property_value = next((k, v) for k, v in node.ref_node.properties.items()
-                                   if k not in (node.properties or ()))
+        try:
+            true_property_value = next((k, v) for k, v in (node.ref_node.properties.items()
+                                                           if node.ref_node.properties else [])
+                                       if k not in (node.properties or ()))
+        except StopIteration:
+            return
```
Comment (danielhers) [inline tupa/oracle.py:201]: Should be `return None` to be more explicit.

**Review by danielhers — CHANGES_REQUESTED** [review]

File: tupa/config.py [hunk 2]
```diff
@@ -22,9 +22,9 @@
 COMPOUND = "compound"
 
 # Required number of edge labels per framework
-NODE_LABELS_NUM = {"amr": 1000, "dm": 1000, "psd": 1000, "eds": 1000, "ucca": 0}
-NODE_PROPERTY_NUM = {"amr": 1000, "dm": 510, "psd": 1000, "eds": 1000, "ucca": 0}
-EDGE_LABELS_NUM = {"amr": 141, "dm": 59, "psd": 90, "eds": 10, "ucca": 15}
+NODE_LABELS_NUM = {"amr": 1000, "dm": 1000, "psd": 1000, "eds": 1000, "ucca": 0, "ptg": 1000}
+NODE_PROPERTY_NUM = {"amr": 1000, "dm": 510, "psd": 1000, "eds": 1000, "ucca": 0, "ptg": 1000}
+EDGE_LABELS_NUM = {"amr": 141, "dm": 59, "psd": 90, "eds": 10, "ucca": 15, "ptg": 150}
```
[the remark this answers is not in the document]
Comment (OfirArviv) [reply tupa/config.py:27]: The default values are enough

File: tupa/states/ref_graph.py [hunk 3]
```diff
@@ -56,3 +78,27 @@ def __init__(self, graph, conllu, framework):
                 node.properties = {prop: resolve(node, value, introduce_placeholders=True)
                                    for prop, value in node.properties.items()}
             node.label = resolve(node, node.label, introduce_placeholders=True)  # Must be after properties in case NAME
+
+
+def find_cycle(graph, plot_graph=False) -> bool:
+    edges_tuple = list(map(lambda x: (x.src, x.tgt), graph.edges))
```
[the remark this answers is not in the document]
Comment (OfirArviv) [reply tupa/states/ref_graph.py]: habit from typescript :) changed it

**Comment by OfirArviv** [comment]
Still failing due to EDS. I'll check later why

**Comment by danielhers** [comment]
Hey @OfirArviv , when you get a chance, could you fix the code to make this PR pass?
https://travis-ci.com/github/danielhers/tupa/jobs/384233352#L806
For EDS, the oracle returns `REDUCE` for an unanchored `poss` node, which violates the constraints; but I think this node is just unanchored in the gold graph so the constraints should allow reducing it.

**Comment by OfirArviv** [comment]
> 
> 
> Hey @OfirArviv , when you get a chance, could you fix the code to make this PR pass?
> https://travis-ci.com/github/danielhers/tupa/jobs/384233352#L806
> For EDS, the oracle returns `REDUCE` for an unanchored `poss` node, which violates the constraints; but I think this node is just unanchored in the gold graph so the constraints should allow reducing it.

Sorry. With all the pressure forgot about it. The question is, why it broke? We didn't change anything related to EDS as far as I can see. Did something changed in the build?

**Comment by danielhers** [comment]
You are right. I finally figured out it's due to a bug in mtool: cfmrp/mtool#84.

**Pull request merged**